File: packages/openmemory-js/src/ops/compress.ts
import { createHash } from "crypto";

export interface CompressionMetrics {
    ogTok: number;
    compTok: number;
    ratio: number;
    saved: number;
    pct: number;
    latency: number;
    algo: string;
    ts: number;
}

export interface CompressionResult {
    og: string;
    comp: string;
    metrics: CompressionMetrics;
    hash: string;
}

export interface CompressionStats {
    total: number;
    ogTok: number;
    compTok: number;
    saved: number;
    avgRatio: number;
    latency: number;
    algos: Record<string, number>;
    updated: number;
}

class MemoryCompressionEngine {
    private stats: CompressionStats = {
        total: 0,
        ogTok: 0,
        compTok: 0,
        saved: 0,
        avgRatio: 0,
        latency: 0,
        algos: {},
        updated: Date.now(),
    };

    private cache = new Map<string, CompressionResult>();
    private readonly MAX = 500;
    private readonly MS = 0.05;

    // Hoist regex patterns to avoid recompilation
    private readonly SEM_FILTERS = [
        /\b(just|really|very|quite|rather|somewhat|somehow)\b/gi,
        /\b(actually|basically|essentially|literally)\b/gi,
        /\b(I think that|I believe that|It seems that|It appears that)\b/gi,
        /\b(in order to)\b/gi,
    ];

    private readonly SEM_REPLACEMENTS: [RegExp, string][] = [
        [/\bat this point in time\b/gi, "now"],
        [/\bdue to the fact that\b/gi, "because"],
        [/\bin the event that\b/gi, "if"],
        [/\bfor the purpose of\b/gi, "to"],
        [/\bin the near future\b/gi, "soon"],
        [/\ba number of\b/gi, "several"],
        [/\bprior to\b/gi, "before"],
        [/\bsubsequent to\b/gi, "after"],
    ];

    private readonly SYN_CONTRACTIONS: [RegExp, string][] = [
        [/\bdo not\b/gi, "don't"],
        [/\bcannot\b/gi, "can't"],
        [/\bwill not\b/gi, "won't"],
        [/\bshould not\b/gi, "shouldn't"],
        [/\bwould not\b/gi, "wouldn't"],
        [/\bit is\b/gi, "it's"],
        [/\bthat is\b/gi, "that's"],
        [/\bwhat is\b/gi, "what's"],
        [/\bwho is\b/gi, "who's"],
        [/\bthere is\b/gi, "there's"],
        [/\bhas been\b/gi, "been"],
        [/\bhave been\b/gi, "been"],
    ];

    private readonly SENTENCE_SPLIT = /[.!?]+\s+/;
    private readonly WHITESPACE_NORM = /\s+/g;
    private readonly ARTICLE_REDUCTION = /\b(the|a|an)\s+(\w+),\s+(the|a|an)\s+/gi;
    private readonly BRACE_COMPRESS = [/\s*{\s*/g, /\s*}\s*/g];
    private readonly PAREN_COMPRESS = [/\s*\(\s*/g, /\s*\)\s*/g];
    private readonly SEMICOLON_COMPRESS = /\s*;\s*/g;

    private readonly AGG_MARKDOWN_CHARS = /[*_~`#]/g;
    private readonly AGG_URL_COMPRESS = /https?:\/\/(www\.)?([^\/\s]+)(\/[^\s]*)?/gi;
    private readonly AGG_ABBREVIATIONS: [RegExp, string][] = [
        [/\bJavaScript\b/gi, "JS"],
        [/\bTypeScript\b/gi, "TS"],
        [/\bPython\b/gi, "Py"],
        [/\bapplication\b/gi, "app"],
        [/\bfunction\b/gi, "fn"],
        [/\bparameter\b/gi, "param"],
        [/\bargument\b/gi, "arg"],
        [/\breturn\b/gi, "ret"],
        [/\bvariable\b/gi, "var"],
        [/\bconstant\b/gi, "const"],
        [/\bdatabase\b/gi, "db"],
        [/\brepository\b/gi, "repo"],
        [/\benvironment\b/gi, "env"],
        [/\bconfiguration\b/gi, "config"],
        [/\bdocumentation\b/gi, "docs"],
    ];
    private readonly AGG_NEWLINE_COMPRESS = /\n{3,}/g;

    private tok(t: string): number {
        if (!t) return 0;
        const w = t.split(/\s+/).length;
        const c = t.length;
        return Math.ceil(c / 4 + w / 2);
    }

    private sem(t: string): string {
        if (!t || t.length < 50) return t;
        let c = t;
        const s = c.split(this.SENTENCE_SPLIT);
        const u = s.filter((x, i, a) => {
            if (i === 0) return true;
            const n = x.toLowerCase().trim();
            const p = a[i - 1].toLowerCase().trim();
            return n !== p;
        });
        c = u.join(". ").trim();
        for (const p of this.SEM_FILTERS) c = c.replace(p, "");
        c = c.replace(this.WHITESPACE_NORM, " ").trim();
        for (const [p, x] of this.SEM_REPLACEMENTS) c = c.replace(p, x);
        return c;
    }

    private syn(t: string): string {
        if (!t || t.length < 30) return t;
        let c = t;
        for (const [p, x] of this.SYN_CONTRACTIONS) c = c.replace(p, x);
        c = c.replace(this.ARTICLE_REDUCTION, "$2, ");
        c = c.replace(this.BRACE_COMPRESS[0], "{").replace(this.BRACE_COMPRESS[1], "}");
        c = c.replace(this.PAREN_COMPRESS[0], "(").replace(this.PAREN_COMPRESS[1], ")");
        c = c.replace(this.SEMICOLON_COMPRESS, ";");
        return c;
    }

    private agg(t: string): string {
        if (!t) return t;
        let c = this.sem(t);
        c = this.syn(c);
        c = c.replace(this.AGG_MARKDOWN_CHARS, "");
        c = c.replace(this.AGG_URL_COMPRESS, "$2");
        for (const [p, x] of this.AGG_ABBREVIATIONS) c = c.replace(p, x);
        c = c.replace(this.AGG_NEWLINE_COMPRESS, "\n\n");
        c = c
            .split("\n")
            .map((l) => l.trim())
            .join("\n");
        return c.trim();
    }

    compress(
        t: string,
        a: "semantic" | "syntactic" | "aggressive" = "semantic",
    ): CompressionResult {
        if (!t) {
            return {
                og: t,
                comp: t,
                metrics: this.empty(a),
                hash: this.hash(t),
            };
        }

        const k = `${a}:${this.hash(t)}`;
        if (this.cache.has(k)) return this.cache.get(k)!;

        const ot = this.tok(t);
        let c: string;

        switch (a) {
            case "semantic":
                c = this.sem(t);
                break;
            case "syntactic":
                c = this.syn(t);
                break;
            case "aggressive":
                c = this.agg(t);
                break;
            default:
                c = t;
        }

        const ct = this.tok(c);
        const sv = ot - ct;
        const r = ct / ot;
        const p = (sv / ot) * 100;
        const l = sv * this.MS;

        const m: CompressionMetrics = {
            ogTok: ot,
            compTok: ct,
            ratio: r,
            saved: sv,
            pct: p,
            latency: l,
            algo: a,
            ts: Date.now(),
        };

        const res: CompressionResult = {
            og: t,
            comp: c,
            metrics: m,
            hash: this.hash(t),
        };
        this.up(m);
        this.store(k, res);
        return res;
    }

    batch(
        ts: string[],
        a: "semantic" | "syntactic" | "aggressive" = "semantic",
    ): CompressionResult[] {
        return ts.map((t) => this.compress(t, a));
    }

    auto(t: string): CompressionResult {
        if (!t || t.length < 50) return this.compress(t, "semantic");
        const code =
            /\b(function|const|let|var|def|class|import|export)\b/.test(t);
        const urls = /https?:\/\//.test(t);
        const verb = t.split(/\s+/).length > 100;
        let a: "semantic" | "syntactic" | "aggressive";
        if (code || urls) a = "aggressive";
        else if (verb) a = "semantic";
        else a = "syntactic";
        return this.compress(t, a);
    }

    getStats(): CompressionStats {
        return { ...this.stats };
    }

    analyze(t: string): Record<string, CompressionMetrics> {
        const r: Record<string, CompressionMetrics> = {};
        for (const a of ["semantic", "syntactic", "aggressive"] as const) {
            const x = this.compress(t, a);
            r[a] = x.metrics;
        }
        return r;
    }

    reset(): void {
        this.stats = {
            total: 0,
            ogTok: 0,
            compTok: 0,
            saved: 0,
            avgRatio: 0,
            latency: 0,
            algos: {},
            updated: Date.now(),
        };
    }

    clear(): void {
        this.cache.clear();
    }

    private empty(a: string): CompressionMetrics {
        return {
            ogTok: 0,
            compTok: 0,
            ratio: 1,
            saved: 0,
            pct: 0,
            latency: 0,
            algo: a,
            ts: Date.now(),
        };
    }

    private hash(t: string): string {
        return createHash("md5").update(t).digest("hex").substring(0, 16);
    }

    private up(m: CompressionMetrics): void {
        this.stats.total++;
        this.stats.ogTok += m.ogTok;
        this.stats.compTok += m.compTok;
        this.stats.saved += m.saved;
        this.stats.latency += m.latency;
        if (this.stats.ogTok > 0)
            this.stats.avgRatio = this.stats.compTok / this.stats.ogTok;
        if (!this.stats.algos[m.algo]) this.stats.algos[m.algo] = 0;
        this.stats.algos[m.algo]++;
        this.stats.updated = Date.now();
    }

    private store(k: string, r: CompressionResult): void {
        if (this.cache.size >= this.MAX) {
            const f = this.cache.keys().next().value;
            if (f) this.cache.delete(f);
        }
        this.cache.set(k, r);
    }
}

export const compressionEngine = new MemoryCompressionEngine();
export { MemoryCompressionEngine };

File: packages/openmemory-py/src/openmemory/ops/compress.py
import re
import time
import hashlib
from typing import Dict, Any, List, Optional

# --- Module-level compiled regex constants ---

_WHITESPACE_RE = re.compile(r"\s+")
_SENTENCE_SPLIT_RE = re.compile(r"[.!?]+\s+")

_FILLER_RES = [
    re.compile(r"\b(just|really|very|quite|rather|somewhat|somehow)\b", re.IGNORECASE),
    re.compile(r"\b(actually|basically|essentially|literally)\b", re.IGNORECASE),
    re.compile(r"\b(I think that|I believe that|It seems that|It appears that)\b", re.IGNORECASE),
    re.compile(r"\b(in order to)\b", re.IGNORECASE),
]

_REPLACEMENT_RES = [
    (re.compile(r"\bat this point in time\b", re.IGNORECASE), "now"),
    (re.compile(r"\bdue to the fact that\b", re.IGNORECASE), "because"),
    (re.compile(r"\bin the event that\b", re.IGNORECASE), "if"),
    (re.compile(r"\bfor the purpose of\b", re.IGNORECASE), "to"),
    (re.compile(r"\bin the near future\b", re.IGNORECASE), "soon"),
    (re.compile(r"\ba number of\b", re.IGNORECASE), "several"),
    (re.compile(r"\bprior to\b", re.IGNORECASE), "before"),
    (re.compile(r"\bsubsequent to\b", re.IGNORECASE), "after"),
]

_CONTRACTION_RES = [
    (re.compile(r"\bdo not\b", re.IGNORECASE), "don't"),
    (re.compile(r"\bcannot\b", re.IGNORECASE), "can't"),
    (re.compile(r"\bwill not\b", re.IGNORECASE), "won't"),
    (re.compile(r"\bshould not\b", re.IGNORECASE), "shouldn't"),
    (re.compile(r"\bwould not\b", re.IGNORECASE), "wouldn't"),
    (re.compile(r"\bit is\b", re.IGNORECASE), "it's"),
    (re.compile(r"\bthat is\b", re.IGNORECASE), "that's"),
    (re.compile(r"\bwhat is\b", re.IGNORECASE), "what's"),
    (re.compile(r"\bwho is\b", re.IGNORECASE), "who's"),
    (re.compile(r"\bthere is\b", re.IGNORECASE), "there's"),
    (re.compile(r"\bhas been\b", re.IGNORECASE), "been"),
    (re.compile(r"\bhave been\b", re.IGNORECASE), "been"),
]

_ARTICLE_STRIP_RE = re.compile(r"\b(the|a|an)\s+(\w+),\s+(the|a|an)\s+", re.IGNORECASE)
_BRACE_OPEN_RE = re.compile(r"\s*{\s*")
_BRACE_CLOSE_RE = re.compile(r"\s*}\s*")
_PAREN_OPEN_RE = re.compile(r"\s*\(\s*")
_PAREN_CLOSE_RE = re.compile(r"\s*\)\s*")
_SEMICOLON_RE = re.compile(r"\s*;\s*")

_MARKDOWN_RE = re.compile(r"[*_~`#]")
_URL_RE = re.compile(r"https?://(www\.)?([^\/\s]+)(/[^\s]*)?", re.IGNORECASE)

_ABBREV_RES = [
    (re.compile(r"\bJavaScript\b", re.IGNORECASE), "JS"),
    (re.compile(r"\bTypeScript\b", re.IGNORECASE), "TS"),
    (re.compile(r"\bPython\b", re.IGNORECASE), "Py"),
    (re.compile(r"\bapplication\b", re.IGNORECASE), "app"),
    (re.compile(r"\bfunction\b", re.IGNORECASE), "fn"),
    (re.compile(r"\bparameter\b", re.IGNORECASE), "param"),
    (re.compile(r"\bargument\b", re.IGNORECASE), "arg"),
    (re.compile(r"\breturn\b", re.IGNORECASE), "ret"),
    (re.compile(r"\bvariable\b", re.IGNORECASE), "var"),
    (re.compile(r"\bconstant\b", re.IGNORECASE), "const"),
    (re.compile(r"\bdatabase\b", re.IGNORECASE), "db"),
    (re.compile(r"\brepository\b", re.IGNORECASE), "repo"),
    (re.compile(r"\benvironment\b", re.IGNORECASE), "env"),
    (re.compile(r"\bconfiguration\b", re.IGNORECASE), "config"),
    (re.compile(r"\bdocumentation\b", re.IGNORECASE), "docs"),
]

_MULTI_NEWLINE_RE = re.compile(r"\n{3,}")
_CODE_DETECT_RE = re.compile(r"\b(function|const|let|var|def|class|import|export)\b")
_URL_DETECT_RE = re.compile(r"https?://")

# --- Engine ---

class MemoryCompressionEngine:
    def __init__(self):
        self.stats = {
            "total": 0,
            "ogTok": 0,
            "compTok": 0,
            "saved": 0,
            "avgRatio": 0,
            "latency": 0,
            "algos": {},
            "updated": int(time.time() * 1000)
        }
        self.cache = {}
        self.MAX = 500
        self.MS = 0.05

    def tok(self, t: str) -> int:
        if not t: return 0
        w = len(_WHITESPACE_RE.split(t.strip()))
        c = len(t)
        return int(c / 4 + w / 2) + 1

    def sem(self, t: str) -> str:
        if not t or len(t) < 50: return t
        c = t
        s = _SENTENCE_SPLIT_RE.split(c)
        u = []
        for i, x in enumerate(s):
            if i == 0:
                u.append(x)
                continue
            n = x.lower().strip()
            p = s[i-1].lower().strip()
            if n != p: u.append(x)

        c = ". ".join(u).strip()
        for pat in _FILLER_RES:
            c = pat.sub("", c)

        c = _WHITESPACE_RE.sub(" ", c).strip()

        for pat, repl in _REPLACEMENT_RES:
            c = pat.sub(repl, c)

        return c

    def syn(self, t: str) -> str:
        if not t or len(t) < 30: return t
        c = t
        for pat, repl in _CONTRACTION_RES:
            c = pat.sub(repl, c)

        c = _ARTICLE_STRIP_RE.sub(r"\2, ", c)
        c = _BRACE_OPEN_RE.sub("{", c)
        c = _BRACE_CLOSE_RE.sub("}", c)
        c = _PAREN_OPEN_RE.sub("(", c)
        c = _PAREN_CLOSE_RE.sub(")", c)
        c = _SEMICOLON_RE.sub(";", c)
        return c

    def agg(self, t: str) -> str:
        if not t: return t
        c = self.sem(t)
        c = self.syn(c)
        c = _MARKDOWN_RE.sub("", c)
        c = _URL_RE.sub(r"\2", c)

        for pat, repl in _ABBREV_RES:
            c = pat.sub(repl, c)

        c = _MULTI_NEWLINE_RE.sub("\n\n", c)
        c = "\n".join([l.strip() for l in c.split("\n")])
        return c.strip()

    def compress(self, t: str, a: str = "semantic") -> Dict[str, Any]:
        if not t:
            return {
                "og": t, "comp": t,
                "metrics": self.empty(a),
                "hash": self.hash(t)
            }

        k = f"{a}:{self.hash(t)}"
        if k in self.cache: return self.cache[k]

        ot = self.tok(t)
        if a == "semantic": c = self.sem(t)
        elif a == "syntactic": c = self.syn(t)
        elif a == "aggressive": c = self.agg(t)
        else: c = t

        ct = self.tok(c)
        sv = ot - ct
        r = ct / ot if ot > 0 else 1
        p = (sv / ot) * 100 if ot > 0 else 0
        l = sv * self.MS

        m = {
            "ogTok": ot, "compTok": ct, "ratio": r, "saved": sv,
            "pct": p, "latency": l, "algo": a, "ts": int(time.time()*1000)
        }
        res = {
            "og": t, "comp": c, "metrics": m, "hash": self.hash(t)
        }
        self.up(m)
        self.store(k, res)
        return res

    def batch(self, ts: List[str], a: str = "semantic") -> List[Dict[str, Any]]:
        return [self.compress(t, a) for t in ts]

    def auto(self, t: str) -> Dict[str, Any]:
        if not t or len(t) < 50: return self.compress(t, "semantic")
        code = bool(_CODE_DETECT_RE.search(t))
        urls = bool(_URL_DETECT_RE.search(t))
        verb = len(t.split()) > 100

        if code or urls: a = "aggressive"
        elif verb: a = "semantic"
        else: a = "syntactic"
        return self.compress(t, a)

    def empty(self, a: str):
        return {
            "ogTok": 0, "compTok": 0, "ratio": 1, "saved": 0,
            "pct": 0, "latency": 0, "algo": a, "ts": int(time.time()*1000)
        }

    def hash(self, t: str) -> str:
        return hashlib.md5(t.encode("utf-8")).hexdigest()[:16]

    def up(self, m):
        self.stats["total"] += 1
        self.stats["ogTok"] += m["ogTok"]
        self.stats["compTok"] += m["compTok"]
        self.stats["saved"] += m["saved"]
        self.stats["latency"] += m["latency"]
        if self.stats["ogTok"] > 0:
            self.stats["avgRatio"] = self.stats["compTok"] / self.stats["ogTok"]

        algo = m["algo"]
        self.stats["algos"][algo] = self.stats["algos"].get(algo, 0) + 1
        self.stats["updated"] = int(time.time()*1000)

    def store(self, k, r):
        if len(self.cache) >= self.MAX:
            first = next(iter(self.cache))
            del self.cache[first]
        self.cache[k] = r

compression_engine = MemoryCompressionEngine()
